feat: add network status component with signal strength indicator

BohdanDzihim opened the pull request:
- Implemented a new NetworkStatusComponent to display the current network signal strength.
- Added HTML structure for the signal bars in network-status.component.html.
- Styled the signal bars with SCSS in network-status.component.scss.
- Integrated logic to update signal strength based on online/offline status and connection speed in network-status.component.ts.
- Added icons for the web app in the manifest.webmanifest file.
- Included new PNG icons for various sizes in the public/icons directory.

BohdanDzihim committed:
feat: add network status component with signal strength indicator

- Implemented a new NetworkStatusComponent to display the current network signal strength.
- Added HTML structure for the signal bars in network-status.component.html.
- Styled the signal bars with SCSS in network-status.component.scss.
- Integrated logic to update signal strength based on online/offline status and connection speed in network-status.component.ts.
- Added icons for the web app in the manifest.webmanifest file.
- Included new PNG icons for various sizes in the public/icons directory.

Protodiv commented on src/index.html:
```diff
@@ -6,10 +6,13 @@
+  <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@300;400;500&amp;display=swap" rel="stylesheet">
   <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
+  <link rel="manifest" href="manifest.webmanifest">
+  <link rel="manifest" href="manifest.webmanifest">
```
Why do you place it two times

Protodiv commented:
The Pokemon service class is named Pokemon instead of following Angular's naming convention of PokemonService.

Protodiv commented:
ngIf is deprecated, 20.0 Use the `@if` block instead.

BohdanDzihim committed:
refactor: rename Pokemon class to PokemonService and update references

BohdanDzihim committed:
refactor: update button visibility condition for reload functionality

Protodiv commented on src/app/features/pokemon-list/pokemon-list.component.ts:
```diff
@@ -58,44 +58,52 @@ export class PokemonList implements OnInit {
   fetchRandomPokemons(): void {
```
This function is too big. Refactor it. It is too hard to read

BohdanDzihim committed:
refactor: update PokemonDetailComponent to use PokemonService and improve error handling in PokemonList component

Protodiv commented on src/app/features/pokemon-list/pokemon-list.component.ts:
```diff
@@ -1,9 +1,9 @@
+import { from, mergeMap, catchError, EMPTY, toArray, forkJoin, Observable } from 'rxjs';
 
 @Component({
   selector: 'app-root',
```
the selector is 'app-root' which should be 'app-pokemon-list' to match the component name and avoid conflicts with the main app component.

Protodiv commented on src/app/features/pokemon-list/pokemon-list.component.ts:
```diff
@@ -58,44 +58,33 @@ export class PokemonList implements OnInit {
   fetchRandomPokemons(): void {
```
Refactor do not destroy your core functionality. 
Break function into smaller pieces

Protodiv commented on src/app/features/pokemon-list/pokemon-list.component.ts:
```diff
@@ -58,44 +58,33 @@ export class PokemonList implements OnInit {
-        );
-
-        forkJoin(detailRequests as Observable<PokemonDetails>[]).subscribe({
+        forkJoin(requests as Observable<PokemonDetails>[]).subscribe({
```
In the fetchRandomPokemons() method, the error handling for the forkJoin observable is missing, which could lead to unhandled errors.

Protodiv commented on src/app/features/pokemon-list/pokemon-list.component.ts:
```diff
@@ -58,44 +58,33 @@ export class PokemonList implements OnInit {
+      error: (err) => {
         this.loading = false;
+        if (!navigator.onLine) {
+          this.errorMessage = "You're offline. PLease check your internet connection.";
```
"PLease" should be "Please".

Protodiv commented on src/app/features/pokemon-list/pokemon-list.component.ts:
```diff
@@ -58,44 +58,33 @@ export class PokemonList implements OnInit {
       next: (data) => {
         const list = data.results;
+        const requests = [
+          ...(list.some(e => e.name.toLowerCase() === 'pikachu') ? [] : [this.pokemonService.getPokemonDetails('https://pokeapi.co/api/v2/pokemon/25/')]),
```
Review what solid 
And delete unnecessary function

Protodiv reviewed (changes requested):
bad code

BohdanDzihim committed:
refactor: streamline fetchRandomPokemons method and enhance Pikachu check in PokemonList component

Protodiv commented on src/app/features/pokemon-list/pokemon-list.component.ts:
```diff
@@ -59,43 +69,38 @@ export class PokemonList implements OnInit {
+      switchMap((data) => {
+        const requests = data.results.map((e: PokemonListResult) => this.pokemonService.getPokemonDetails(e.url));
+        if (!this.hasPikachu(data.results)) {
+          requests.push(this.pokemonService.getPokemonDetailsByName('pikachu'));
```
Magic Value

Protodiv commented on src/app/features/pokemon-list/pokemon-list.component.ts:
```diff
@@ -59,43 +69,38 @@ export class PokemonList implements OnInit {
-          this.errorMessage = 'Failed to load Pokemon list.';
+    this.pokemonService.getPokemons(limit, offset).pipe(
+      switchMap((data) => {
+        const requests = data.results.map((e: PokemonListResult) => this.pokemonService.getPokemonDetails(e.url));
```
There is no need for getPokemonDetails at all

Protodiv commented on src/app/features/pokemon-list/pokemon-list.component.ts:
```diff
@@ -59,43 +69,38 @@ export class PokemonList implements OnInit {
+        return forkJoin(requests);
+      })
+    ).subscribe({
+      next: (results) => {
```
Move it to a separate function

Protodiv commented on src/app/features/pokemon-list/pokemon-list.component.ts:
```diff
@@ -59,43 +69,38 @@ export class PokemonList implements OnInit {
         this.loading = false;
+      },
+      
+      error: (err) => {
```
Move it to a separate function

BohdanDzihim committed:
refactor: enhance error handling and streamline fetching of Pokémon details in PokemonList component

2 private functions were created in PokemonList component to follow SOLID principles. each method should handle a single responsibility as in handleNext and handleError, where its sole responsibility is to handle errors and results. fetchRandomPokemons function should only handle fetching pokemons, which it does

Protodiv commented:
One left

BohdanDzihim committed:
refactor: replace direct Pikachu fetch with dedicated getPikachu method in PokemonList component

Protodiv commented on src/app/core/services/pokemon.service.ts:
```diff
@@ -15,14 +15,14 @@ export class Pokemon {
 
-  getPokemonDetails(url: string): Observable<PokemonDetails> {
-    return this.http.get<PokemonDetails>(url);
+  getPikachu(): Observable<PokemonDetails> {
```
What is this (This is a bad function 
 it's only called for pickachu 

and even with this code can be smarter 

getPikachu(): Observable<PokemonDetails> {
return getPokemonDetailsByName('pikachu');
}

but this is not what I meant 

this function must be deleted

Protodiv commented on src/app/features/pokemon-list/pokemon-list.component.ts:
```diff
@@ -55,48 +65,47 @@ export class PokemonList implements OnInit {
+        
+        if (!this.hasPikachu(data.results)) {
+          requests.push(this.pokemonService.getPikachu());
         }
```
I was expecting you to create a static variable that host value for the default pokemon 
DEFAULT_POKEMON = 'pikachu'
And place it in pokemon.service
and write down for me why there and not in component

BohdanDzihim committed:
refactor: replace direct Pikachu fetch with getPikachu method using DEFAULT_POKEMON constant

Protodiv commented on src/app/core/services/pokemon.service.ts:
```diff
@@ -6,23 +6,24 @@ import { Observable } from 'rxjs';
 
-  getPokemonDetailsByName(name: string): Observable<PokemonDetails> {
-    return this.http.get<PokemonDetails>(`${Pokemon.BASE_URL}${Pokemon.ENDPOINTS.POKEMON}/${name}`);
+  getPikachu(): Observable<PokemonDetails> {
```
Delete unnecessary fun

use getPokemonDetailsByName

Protodiv commented on src/app/features/pokemon-list/pokemon-list.component.ts:
```diff
@@ -41,6 +41,16 @@ export class PokemonList implements OnInit {
   }
 
+  hasPikachu(list: PokemonListResult[]): boolean {
+    return list.some(pokemon => pokemon.name?.toLowerCase() === 'pikachu');
```
use public variable

Protodiv commented on src/app/features/pokemon-list/pokemon-list.component.ts:
```diff
@@ -41,6 +41,16 @@ export class PokemonList implements OnInit {
     this.pokemons = sorted;
   }
 
+  hasPikachu(list: PokemonListResult[]): boolean {
```
logically rename fun

BohdanDzihim committed:
refactor: replace hasPikachu method with hasDefaultPokemon and update Pikachu fetch logic in PokemonList component

Protodiv reviewed (approved):
Good Job
